### PR TITLE
fix(cli): broaden write-through cache envelope keys to match sync

### LINKED ISSUE #1856: generator: writeThroughCache envelope walker recognises fewer keys than sync.go.tmpl (asymmetry)
> Filed by a manual retro after generating an out-of-catalog SofaScore CLI (https://www.sofascore.com — sniffed; not for public-library publish). Surfaces a template asymmetry between `sync.go.tmpl` and `data_source.go.tmpl` that bites any API whose list endpoints wrap their array in a resource-named property instead of `results` / `data` / `items`.

## Summary

`data_source.go.tmpl`'s `writeThroughCache` / `writeReadResultToStore` only descend into envelope keys `results`, `data`, `items`. `sync.go.tmpl`'s `extractPageItems` recognises a broader set (`data`, `results`, `items`, `records`, `nodes`, `entries` + PascalCase variants) AND scans every envelope key for a single-array shape as a fallback. Any list endpoint that wraps its array in something the auto-cache doesn't know about (e.g. `{"events":[...]}`, `{"standings":[...]}`, `{"incidents":[...]}`) gets its whole envelope passed to `UpsertBatch` as one item, which then fails ID extraction and emits the misleading `warning: N/M <resource> items skipped (no extractable ID field found)` (see also #1224).

The asymmetry means: a `sync` run can succeed on the same endpoint where `auto`-mode read commands silently fail to cache. The two code paths walk the same wire shape with different rules.

## Symptom

Reproduced on every SofaScore list endpoint (each uses a resource-named wrapper):

```bash
$ sofascore-pp-cli sport events list-live --sport football --json
warning: 1/1 events items skipped (no extractable ID field found)
{
  "meta": {"source": "live"},
  "results": {
    "events": [
      {"id": 16184005, "awayTeam": {...}, "homeTeam": {...}, ...},
      ...
    ]
  }
}
```

212 live football events in the response, each with a top-level `id` — but `writeThroughCache` saw `{events: [...]}` as a single-key envelope with no recognised wrapper, fell through to the "treat as detail object" branch, and asked `UpsertBatch` to find an `id` on the wrapper itself. Result: `1/1 skipped`, no rows landed, FTS search + analytics on cached SofaScore data was empty.

After patching the printed CLI to add `events`, `statistics`, `incidents`, `standings`, `players`, `uniqueTournamentSeasons` to both the `listEnvelopeMetadataKeys` set and the two inline `[]string{"results", "data", "items"}` loops in `internal/cli/data_source.go`, the same call cached **212 events** and a follow-up `search Trabzonspor --data-source local` matched real rows.

## Root cause

Two templates walk wrapped responses with different rule sets:

**`internal/generator/templates/data_source.go.tmpl`** (`writeThroughCache`, around the printed-CLI file's lines 207-266):

```go
for _, key := range []string{"results", "data", "items"} {
    if raw, ok := envelope[key]; ok {
        var arr []json.RawMessage
        if json.Unmarshal(raw, &arr) == nil && len(arr) > 0 {
            items = arr
            break
        }
    }
}
```

Hardcoded to three keys. If none match, `looksLikeListEnvelope` stays false and the whole envelope is upserted as one item.

**`internal/generator/templates/sync.go.tmpl`** (`extractPageItems`, around line 1568 + the multi-key scan around 1236-1255):

```go
var pageItemKeys = []string{
    "data", "results", "items", "records", "nodes", "entries",
    "Data", "Results", "Items", "Records", "Nodes", "Entries",
}
// ... and, when none of those match, a scan-all-keys-for-single-array fallback.
```

Broader set + a fallback that finds resource-named wrappers like `{markets:[...]}`. That's why sync can resolve the same envelope that the auto-cache cannot.

The proposed `x-list-fields` extension in `docs/plans/2026-05-07-002-feat-cli-multi-array-envelope-x-list-fields-plan.md` (status: `proposed`) would solve this cleanly by letting the generator emit a per-resource wrapper-field map that both templates consult. But until that plan lands, the two walkers stay out of step.

## Cross-API evidence

Same root-cause family across several open issues:

- **#1658** — `sync.go.tmpl` missed `features` / `Features` for GeoJSON `FeatureCollection`. Identical shape, sister template. The proposed fix there (append to `pageItemKeys`) only patches sync; `data_source.go.tmpl`'s walker still ignores `features`.
- **#1143** (closed) — `--select` failed against list envelopes for the same root cause family: helper code only descended into `results`/`data`/`items`-named arrays.
- **#894** (closed) — provenance envelope double-wrapped against single-key resource envelopes.
- **#1224** — the `"items skipped"` warning itself is misleading; this issue is one of its top symptoms.

APIs known to be affected:
- **SofaScore** (this run) — `{events: [...]}`, `{standings: [...]}`, `{incidents: [...]}`, `{statistics: [...]}`, `{players: [...]}`, `{uniqueTournamentSeasons: [...]}`. https://www.sofascore.com .
- **USGS Earthquakes / NOAA Weather / any GeoJSON** — `{features: [...]}` (see #1658).
- **Kalshi** — `{markets: [...]}` (already worked around in published library; was the motivation for the deferred `x-list-fields` plan).
- Any vendor that wraps lists in a plural resource name (common in JSON-API-influenced designs).

## Proposed fix

Two viable options, in order of preference:

**Option A — Align `data_source.go.tmpl`'s walker with `sync.go.tmpl`.** Replace the two hardcoded `[]string{"results", "data", "items"}` loops with a shared list mirroring `pageItemKeys` (lowercase + PascalCase), and add the single-array scan-all-keys fallback that `extractPageItems` already uses. Then both walkers handle `features`/`markets`/`events`/etc. uniformly. Risk: a detail response with one named field that happens to be a non-empty array (e.g. `{id: "X", tags: ["a", "b"]}`) could get false-positive'd as a list. Mitigation: the existing `looksLikeListEnvelope` guard already requires every other key to be a known list-envelope-metadata key — keep that guard.

**Option B — Land `x-list-fields` per the proposed plan.** Generator emits a per-resource wrapper-field map; both `data_source.go.tmpl` and `sync.go.tmpl` consult it before falling back. Cleaner long-term but bigger surface (auto-emission heuristic, store schema `kind` column, parser/spec extension contract). The plan already documents this.

Either way, the asymmetry between sync and auto-cache should disappear: both walkers reading the same shape should agree.

## Tests

- **Positive (auto-cache):** Generate a CLI from a spec whose list endpoint returns `{"events":[{"id":1,...},{"id":2,...}]}`. Run the read command in `auto` mode, then `analytics --type events --json`: assert the cache holds N rows, not 1.
- **Positive (sync parity):** Run `sync` against the same endpoint; assert the cache count matches the auto-cache count from the previous test.
- **Negative (detail-with-array regression):** Generate against a spec whose detail endpoint returns `{"id":"X","tags":["a","b"]}`. Assert auto-cache stores 1 row, not 2 tag-rows (i.e. the broadened walker does not misread detail responses with arrays as lists).
- **Goldens:** Add a fixture spec with a `{<resource-named-key>: [...]}` response under `testdata/golden/fixtures/`, and lock in the generated `data_source.go` shape via the golden harness.

## Workaround until fixed

Hand-patch the printed CLI's `internal/cli/data_source.go` per regen, extending both the `listEnvelopeMetadataKeys` set and the inline envelope-key loops with the resource-named wrappers the spec actually uses (e.g. `events`, `standings`, ...). This is what we did locally for SofaScore. Not portable across CLIs; this is exactly the kind of per-CLI carry-cost the machine should eat instead.

## Relation to the AGENTS.md "Machine vs Printed CLI" rule

This is squarely a machine change: any future spec with a resource-named list wrapper hits the same bug. We deliberately did the printed-CLI fix locally rather than push a template change without an issue, per the ownership rules.

### PR DESCRIPTION
**Pull request by vinnyp**
## Intent

Issue: Fixes #1856

`writeThroughCache` (`data_source.go.tmpl`) only descended into envelope keys `results`/`data`/`items`, while `sync.go.tmpl`'s `extractPageItems` recognises a broader set plus a single-array fallback. An API that wraps its list in a resource-named key (e.g. `{"events":[...]}`) cached fine under `sync` but the auto write-through upserted the whole envelope as one row (`warning: N/M items skipped (no extractable ID field found)`). The two paths walked the same wire shape with different rules.

## Approach

Bring `writeThroughCache` to parity with `extractPageItems`: use one local key slice matching sync's `pageItemKeys` (`data,results,items,records,nodes,entries` + PascalCase) in both envelope loops, and add the single-array fallback — if no known key matches, use a top-level array iff exactly one key maps to a non-empty array of objects. The existing detail-object guard (a detail object with an empty wrapper-named field still caches as one row) is preserved.

## Scope

Primary area: generator (`internal/generator/templates/data_source.go.tmpl`). Machine change. `sync.go.tmpl` unchanged.

## Risk

Low. Broadens recognition and adds a conservative fallback (object-shaped elements, exactly-one-array); best-effort silent behavior intact.

## Output Contract

Generated `internal/cli/data_source.go` gains the broadened key slice + fallback. Golden fixture `generate-golden-api` (data_source.go) regenerated.

## Verification

- `go build`, `go vet ./...`, `golangci-lint run ./...` (0 issues), `go test -short ./...` — all pass (incl. new `TestWriteThroughCacheEnvelopeExtractionParity`: `{"events":[...]}` caches, `{"results":[...]}` caches, detail object stays single-row)
- `scripts/golden.sh verify` → 1 intentional case; `update` applied; diff inspected

## AI / Automation Disclosure

- [ ] No AI or automation was used
- [ ] Human-reviewed: AI or automation was used, and a human reviewed the work for intent, fit, and obvious issues before submission
- [x] AI-reviewed only: an AI agent reviewed the work, but no human reviewed it before submission
- [ ] Fully automated: generated and submitted without human review for this specific change

### CODE REVIEW

File: internal/generator/templates/data_source.go.tmpl [hunk 1]
```diff
@@ -201,6 +201,11 @@ func writeThroughCache(ctx context.Context, resourceType string, data json.RawMe
 	}
 	defer db.Close()
 
+	pageItemKeys := []string{
+		"data", "results", "items", "records", "nodes", "entries",
+		"Data", "Results", "Items", "Records", "Nodes", "Entries",
+	}
```
[the remark this answers is not in the document]
Comment (vinnyp) [reply internal/generator/templates/data_source.go.tmpl:210]: Applied — added `features`/`Features` so the local slice matches `sync.go.tmpl`'s `pageItemKeys` exactly (14 entries). Regenerated the data_source.go golden. Pushed.

File: internal/generator/writethrough_envelope_test.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		t.Fatalf("decode cached order: %v", err)
+	}
+	if got, ok := order["id"].(string); !ok || got != "x" {
+		t.Fatalf("expected cached detail object id to be x, got %#v", order["id"])
+	}
+}
+`), 0o644))
+
+	runGoCommand(t, outputDir, "test", "./internal/cli", "-run", "TestWriteThroughCacheEnvelopeExtractionParity", "-count=1")
+}
```
[the remark this answers is not in the document]
Comment (vinnyp) [reply internal/generator/writethrough_envelope_test.go:115]: The fallback (`{"events":[...]}`), the `results` fast path, and the detail-object guard are covered; the broadened keys (`records`/`nodes`/`entries`/`features`) are now exact-parity with sync's already-tested `extractPageItems`, which is the main guarantee. Agreed the empty-page edge (`{"records":[],"next_cursor":...}`) and per-key cases would tighten it further — happy to add them as a follow-up; the detail-object guard test already exercises the empty-wrapper-array path.

File: internal/generator/templates/data_source.go.tmpl [hunk 3]
```diff
@@ -219,6 +224,28 @@ func writeThroughCache(ctx context.Context, resourceType string, data json.RawMe
...
+					if json.Unmarshal(candidate[0], &obj) != nil {
+						continue
+					}
+					arrayItems = candidate
+					arrayCount++
+				}
+				if arrayCount == 1 {
+					items = arrayItems
+				}
+			}
```
[the remark this answers is not in the document]
Comment (vinnyp) [reply internal/generator/templates/data_source.go.tmpl:270]: Good catch, fixed. The single-array fallback now only fires when every *sibling* top-level key is a known pagination/metadata key (`listEnvelopeMetadataKeys`). So `{"events":[...]}` and `{"events":[...],"cursor":"x"}` still cache as a list, but a detail object like `{"id":"order-1","line_items":[...]}` — where `id`/`status` aren't metadata keys — falls through to single-row detail caching. Added a regression case with a multi-element sub-array (so a misclassification would yield 2 rows instead of 1) to the behavioral test. Pushed.

**Comment by tmchow** [comment]
@vinnyp while you're at it, can you also fix the `listEnvelopeMetadataKeys` issue greptile flags?

**Comment by vinnyp** [comment]
Done — pushed 7cc4d62a. Added the remaining wrapper keys (`records`/`nodes`/`entries`/`features` plus the PascalCase forms) to `listEnvelopeMetadataKeys` so it stays in sync with `pageItemKeys`. Now an empty page like `{"records":[],"cursor":"abc"}` is recognized as an empty list envelope instead of being cached as a spurious detail row. Added a regression case to the envelope-parity test and regenerated the `data_source.go` golden.
